clusterloader2/cmd/clusterloader.go

```diff
@@ -124,6 +124,7 @@ func validateClusterFlags() *errors.ErrorList {
 }
 
 func initFlags() {
+	klog.InitFlags(nil)
```

Comment:
This is a bit weird - without this line the failure is "--v" flag unknown, with this line it is duplicate failure...

Reply:
With this line:
```
./clusterloader flag redefined: v
panic: ./clusterloader flag redefined: v

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc00019e0e0, {0x2bd2ae0, 0x413b730}, {0x2bbca60, 0x1}, {0x25dc117, 0x22})
	/usr/local/go/src/flag/flag.go:1028 +0x37d
k8s.io/klog/v2.InitFlags.func1(0xc00022c960?)
	/home/prow/go/pkg/mod/k8s.io/klog/v2@v2.130.1/klog.go:447 +0x31
flag.(*FlagSet).VisitAll(0x7fa1040fdb18?, 0xc0009c5c50)
	/usr/local/go/src/flag/flag.go:458 +0x42
k8s.io/klog/v2.InitFlags(0xc0008cba08?)
	/home/prow/go/pkg/mod/k8s.io/klog/v2@v2.130.1/klog.go:446 +0x3c
main.initFlags()
	/home/prow/go/src/k8s.io/perf-tests/clusterloader2/cmd/clusterloader.go:127 +0x19
main.main()
	/home/prow/go/src/k8s.io/perf-tests/clusterloader2/cmd/clusterloader.go:261 +0x45
```

Reply:
Without this line:
```
 F1213 12:42:20.574783   21458 clusterloader.go:262] Flag parse failed: unknown flag: --v
```

Reply:
In which package do you have this problem?

Reply:
Never mind, it's `./clusterloader`.

Reply:
The command line of clusterloader gets set up in clusterloader2/pkg/flags/flags.go:
```
    pflag.CommandLine = pflag.NewFlagSet(os.Args[0], pflag.ContinueOnError)
    // klog.InitFlags(nil)
    pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
```

At the time of that `AddGoFlagSet`, `-v` is not in `flag.CommandLine` yet because the code which now adds it gets invoked later.

The solution is to remove this entire `init` and do the flag set in a more controlled fashion in `main()`.

Reply:
```patch
diff --git a/clusterloader2/cmd/clusterloader.go b/clusterloader2/cmd/clusterloader.go
index f6eab3ec1..810b5ff82 100644
--- a/clusterloader2/cmd/clusterloader.go
+++ b/clusterloader2/cmd/clusterloader.go
@@ -17,6 +17,7 @@ limitations under the License.
 package main
 
 import (
+	"flag"
 	"fmt"
 	"io/ioutil"
 	"net/http"
@@ -25,6 +26,7 @@ import (
 	"path"
 	"time"
 
+	"github.com/spf13/pflag"
 	"gopkg.in/yaml.v2"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/client-go/kubernetes"
@@ -124,6 +126,8 @@ func validateClusterFlags() *errors.ErrorList {
 }
 
 func initFlags() {
+	pflag.CommandLine = pflag.NewFlagSet(os.Args[0], pflag.ContinueOnError)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
 	flags.StringVar(&clusterLoaderConfig.ReportDir, "report-dir", "", "Path to the directory where the reports should be saved. Default is empty, which cause reports being written to standard output.")
 	// TODO(https://github.com/kubernetes/perf-tests/issues/641): Remove testconfig and testoverrides flags when test suite is fully supported.
 	flags.StringArrayVar(&testConfigPaths, "testconfig", []string{}, "Paths to the test config files")
diff --git a/clusterloader2/pkg/flags/flags.go b/clusterloader2/pkg/flags/flags.go
index 9dc866a41..0288d2b14 100644
--- a/clusterloader2/pkg/flags/flags.go
+++ b/clusterloader2/pkg/flags/flags.go
@@ -17,7 +17,6 @@ limitations under the License.
 package flags
 
 import (
-	"flag"
 	"fmt"
 	"os"
 	"strconv"
@@ -27,11 +26,6 @@ import (
 	"github.com/spf13/pflag"
 )
 
-func init() {
-	pflag.CommandLine = pflag.NewFlagSet(os.Args[0], pflag.ContinueOnError)
-	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
-}
-
 var flags []flagFunc
 
 // StringVar creates string flag with given parameters.
```

```console
perf-tests/clusterloader2$ go run ./cmd --help
...
      --testoverrides stringArray                     Paths to the config overrides file. The latter overrides take precedence over changes in former files.
      --testsuite string                              Path to the test suite config file
  -v, --v valueChain                                  number for the log level verbosity
      --vmodule valueChain                            comma-separated list of pattern=N settings for file-filtered logging (only works for the default text log format)
  F1213 15:15:29.635877   74280 clusterloader.go:266] Flag parse failed: pflag: help requested
exit status 1
```

Reply:
FWIW, the conflict occurs because you are indirectly depending on `k8s.io/component-base/logs/testinit`:
```console
$ go run ./cmd --help
/tmp/go-build2255311573/b001/exe/cmd flag redefined: v
panic: /tmp/go-build2255311573/b001/exe/cmd flag redefined: v

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc0001520e0, {0x2bd1960, 0x413a6b0}, {0x2bbb8c8, 0x1}, {0x25db079, 0x22})
	/nvme/gopath/go-1.23.0/src/flag/flag.go:1028 +0x37d
k8s.io/component-base/logs.AddGoFlags.func1(0xc0003ffe00?)
	/nvme/gopath/pkg/mod/k8s.io/component-base@v0.30.4/logs/logs.go:135 +0x130
flag.(*FlagSet).VisitAll(0x0?, 0xc0004addc0)
	/nvme/gopath/go-1.23.0/src/flag/flag.go:458 +0x42
k8s.io/component-base/logs.AddGoFlags(0xc0001520e0, {0x0, 0x0, 0x2d?})
	/nvme/gopath/pkg/mod/k8s.io/component-base@v0.30.4/logs/logs.go:121 +0x91
k8s.io/component-base/logs/testinit.init.0()
	/nvme/gopath/pkg/mod/k8s.io/component-base@v0.30.4/logs/testinit/testinit.go:31 +0x25
exit status 2

$ go mod why k8s.io/component-base/logs/testinit
# k8s.io/component-base/logs/testinit
k8s.io/perf-tests/clusterloader2/pkg/measurement/common/metrics
k8s.io/kubernetes/test/e2e/framework
k8s.io/component-base/logs/testinit
```

That's when doing the `klog.Init` in `pkg/flags.init`, that then runs before `testinit`. So basically two inits fight over who gets to register the flag.

That `k8s.io/component-base/logs/testinit` is supposed for tests, so it seemed okay to import it in `k8s.io/kubernetes/test/e2e/framework`. Why does the `clusterloader2` *command* depend on the `test/e2e/framework`?

Reply:
For historical reasons I guess - it was the simplest thing to do. We rely on some utility functions from test/e2e/framework. We can probably copy those and remove this dependency...

Reply:
Primarily this one: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/ssh/ssh.go